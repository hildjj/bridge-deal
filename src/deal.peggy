{{
import {DealRules} from './ast.js'
import {Direction, Vuln} from './card.js'

function suitOps(op, num, suits) {
  let ret = [];
  for (const s of suits) {
    ret.push(`(dir.${s}.length ${op} ${num})`);
  }
  return ret;
}
}}

{
const rules = new DealRules();
}

commands
  = (_ command? (_ comment)?)|.., CRLF| CRLF? { return rules.toString(); }

command
  = dealer
  / vuln
  / hand
  / setVar
  / bid
  / rule
  / comment

dealer
  = "dealer" __ d:dir { rules.dealer = d; }

hand
  = d:dir { rules.dir = d }

setVar
  = "$" id:identifier _ "=" _ desc:description {
    rules.add({[id]: desc});
  }
  / "%" id:identifier _ "=" _ nums:($[0-9]+)|1.., _ "," _| {
    rules.add({[id]: nums});
  }

dir
  = "north"i { return Direction.NORTH; }
  / "south"i { return Direction.SOUTH; }
  / "east"i { return Direction.EAST; }
  / "west"i { return Direction.WEST; }

vuln
  = "vuln" __ vuln:vulnerability { rules.vuln = vuln; }

vulnerability
  = "all"i { return Vuln.ALL; }
  / "both"i { return Vuln.ALL; }
  / "NS"i { return Vuln.NS; }
  / "EW"i { return Vuln.EW; }
  / "none"i { return Vuln.NONE; }
  / "neither"i { return Vuln.NONE; }
  / "random"i { return undefined; }

bid
 = "bid" __ bid:description {
    rules.bids.push(bid);
  }

suit
  = "C"i { return 'clubs'; }
  / "D"i { return 'diamonds'; }
  / "H"i { return 'hearts'; }
  / "S"i { return 'spades'; }

comment
  = "//" $commentChar*

commentChar
  = !CRLF .

description
  = desc:$normalChar+ { return desc.trim(); }

normalChar
  = !CRLF !"//" .

rule
  = not:(@"!"|0..2| _)? parts:part|1.., "," _| {
    rules.add(parts.join(' && '), not?.length ?? 0);
  }

// Return the "not" versions
part
  = holding
  / weight
  / compare
  / distribution
  / singleton
  / void
  / min:hcp "-" max:hcp { return `dir.range(${min}, ${max})`; }
  / p:hcp more:"+"? {
      if (more) {
        return `(dir.points >= ${p})`;
      } else {
        return `(dir.points === ${p})`
      }
    }
  / "balanced5" { return 'dir.balanced()'; }
  / "balanced" { return 'dir.balancedNoM()'; }
  / cards

hcp
  = @n:$([0-9] [0-9]?) &{ return parseInt(n, 10) <= 37; }

weight
  = suit:suit "%" id:ref _ op:op _ num:$[0-9]+ {
    return `(Deal.weight(dir.${suit}, ${id}) ${op} ${num})`;
  }

ref
  = id:identifier {
    if (!rules.isVar(id)) {
      error(`Unknown variable: "${id}"`);
    }
    return id;
  }

compare
  = a:suit _ op:op _ b:suit {
    return `(dir.${a}.length ${op} dir.${b}.length)`;
  }

op
  = $([<>] "=")
  / [<>]
  / "=" { return "==="; }

holding
  = holdingPlus
  / holdingLess
  / holdingRange

holdingPlus
  = num:[0-9] plus:"+"? suits:(suitSet / suit+) {
      const op = plus ? ">=" : "==="
      return '(' + suitOps(op, num, suits).join(' || ') + ')';
    }

holdingLess
  = op:$("<" "="?) _ num:[0-9] suits:(suitSet / suit+) {
      return suitOps(op, num, suits).join(' && ');
    }

holdingRange
  = a:[0-9] "-" b:[0-9] suits:(suitSet / suit+) {
      return suitOps(">=", a, suits).join(' && ') + ' && ' +
        suitOps("<=", b, suits).join(' && ');
    }

suitSet
  = "X"i { return ['clubs', 'diamonds', 'hearts', 'spades'] }
  / "M"  { return ['hearts', 'spades'] }
  / "m"  { return ['clubs', 'diamonds'] }
  / "P" { return ['diamonds', 'spades'] }
  / "R" { return ['clubs', 'hearts'] }
  / "r" { return ['diamonds', 'hearts'] }
  / "b" { return ['clubs', 'spades'] }

distribution
  = any:"any"? _ dist:[0-9]|4| {
    if (any) {
      return `dir.isShapeAny(${dist.join(', ')})`
    }
    return `dir.isShape(${dist.join(', ')})`
  }

singleton
  = "singleton" {
    return `dir.hasSingleton()`;
  }

void
  = "void" {
    return `dir.hasVoid()`;
  }

cards
  = sc:suitCards|1..4, __| {
    return '(' + sc.join(' && ') + ')';
  }

suitCards
  = cards:$(suit card+) {
    return `dir.hasCards('${cards}')`;
  }

card
  = [2-9TJQKA]

__
  = WS+ {}

_
  = WS* {}

WS
  = [ \t] {}

CRLF
  = "\r\n" {}
  / "\n" {}

identifier
  = $(XID_Start XID_Continue*)

XID_Start
  = c:ValidSourceCharacter &{ return /\p{XID_Start}/u.test(c) }

XID_Continue
  = c:ValidSourceCharacter &{ return /\p{XID_Continue}/u.test(c) }

ValidSourceCharacter
  = _SourceCharacterLow
  / _SurrogatePair

// Not surrogates
_SourceCharacterLow
  = [\u0000-\uD7FF\uE000-\uFFFF]

_SurrogatePair
  = $( [\uD800-\uDBFF][\uDC00-\uDFFF] )
