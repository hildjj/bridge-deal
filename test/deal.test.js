import {default as assert, equal, throws} from 'node:assert/strict';
import {parse} from '../docs/js/deal.js';
import snap from 'snappy-snaps';
import {test} from 'node:test';
import {testPeggy} from '@peggyjs/coverage';

async function snapParse(src) {
  try {
    const filter = parse(src, {
      grammarSource: import.meta.url,
    });
    const expected = await snap(src.replaceAll('\r', ''), filter);
    equal(filter, expected);
  } catch (e) {
    if (e.format) {
      e.message = e.format([{source: import.meta.url, text: src}]);
    }
    throw e;
  }
}

test('peggyCoverage', async() => {
  await testPeggy(new URL('../docs/js/deal.js', import.meta.url), [
    {
      validInput: '',
      validResult: "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
      invalidInput: 'aaa',
    },
    {invalidInput: 'north\naaa'},
    {invalidInput: '/'},
    {invalidInput: 'bid:'},
    {invalidInput: 'north\nbid:'},
  ]);
});

test('parse', async() => {
  for (const src of [
    'dealer north',
    'dealer south',
    'dealer eAst',
    'dealer west',
    'vuln all',
    'vuln BOTH',
    'vuln ns',
    'vuln ew',
    'vuln none',
    'vuln neither',
    'vuln random',
    'north',
    'south',
    'east',
    'west',
    'north\nsouth // foo',
    'north\n$suit = S',
    'south\n%points = 4, 3, 2, 1',
    'bid P',
    'bid 1C!: 16+, artificial',
    'bid 1C!://\n',
    'bid 1C!:// \n',
    'bid 1C!:// ',
    'north\n  !22-23, balanced',
    'north\n  !22-23, balanced5',
    'north\n  8',
    'north\n  13-15',
    'north\n  13+',
    'north\n  13',
    'north\n  37',
    'north\n  4CD',
    'north\n  4HS',
    'north\n  4+M',
    'north\n  <6R',
    'north\n  <5r',
    'north\n  <5b',
    'north\n  <=6P',
    'north\n  4-6S',
    'north\n  4X',
    'north\n  4m',
    'north\n  S > H',
    'north\n  S >= H',
    'north\n  S = H',
    'north\n  S < H',
    'north\n  S <= H',
    'north\n  %POINTS = 40, 30, 2, 1\n  S%POINTS > 3',
    'north\n  %POINTS = 40, 30, 2, 1\n  S%POINTS > 30',
    'north\n  4441',
    'north\n  any 4441',
    'north\r\n  CAK',
    'north\n  $ðž‹„ = H',
    'north\n  !!11-13\n  4D',
    'north\n  void',
    'north\n  singleton',
    '',
    '\n',
    '//',
    '// ',
    '//\n',
    '//a',
    '//a\n',
    '//ab',
    '//ab\n',
    '//abc',
    '//abc\n',
    '//abcd',
    '//abcd\n',
    '//abcde',
    '//abcde\n',
    '//abcdef',
    '//abcdef\n',
    '//abcdefg',
    '//abcdefg\n',
  ]) {
    await snapParse(src);
  }

  assert(parse('', {startRule: 'commands'}));
  assert(parse('', {peg$library: true}));
});

test('parse errors', () => {
  try {
    parse('foo', {
      grammarSource: 'testing',
    });
  } catch (er) {
    equal(typeof er.format, 'function');
    snap('format error', er.format([{
      source: 'testing',
      text: 'foo',
    }]));
  }
  for (const src of [
    'foo',
    'nor',
    'north\n  C%foo = 2',
    'north\n  8,',
    'north\n  8a',
    'north\n  8-a',
    'north\n  8-9a',
    'north\n  38',
    'north\n  <',
    'north\n  <=',
    'north\n  <=6I',
    'north\n  <6HI',
    'north\n  H < I',
    'north\n  C',
    'north\n  CAI',
    'north\n  CA ',
    'north\n  CA DJ HT S9 C4',
    'north\n  $',
    'north\n  $\n',
    'north\n  $ \n',
    'north\n  $ ',
    'north\n  $}',
    'north\n  $} = S',
    'north\n  $a} = S',
    'north\n  $a\ud800 = S',
    'north\n  $a =',
    'north\n  $a =\n',
    'north\n  %',
    'north\n  %\n',
    'north\n  % \n',
    'north\n  % ',
    'north\n  %}',
    'north\n  %} = 1',
    'north\n  %a} = 2',
    'north\n  %a\ud800 = 3',
    'north\n  %a =',
    'north\n  %a =\n',
    'north\n  %a = b\n',
    'north\n  %a = 1,\n',
    'north\n  %a = 1,b\n',
    'north\n  C%}',
    'north\n  C%a',
    'north\n  C%a < 4',
    'north\n  %a=4,2\nC%a .',
    'north\n  %a=4,2\nC%a < b',
    'north\n  %a=4,2\nC%a < 1b',
    'north ',
    'dealer',
    'dealer ',
    'dealer\n',
    'vuln',
    'vuln\n',
    'vuln ',
    'bid',
    'bid ',
    'bid\n',
    '\n I',
  ]) {
    throws(() => parse(src, {
      grammarSource: 'test',
    }), er => {
      if (er.format) {
        assert(er.format([{source: 'test', text: src}]));
        return true;
      }
      return false;
    }, `"${src}"`);
  }

  throws(() => parse('', {startRule: 'ref'}));
  throws(() => parse('', {startRule: 'BAD RULE NAME'}));
});
