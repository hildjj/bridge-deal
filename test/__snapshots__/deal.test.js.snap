// Data Snap v1

exports[``] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`// `] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`// foo`] = {
  "data": "let dir = null;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`// foo
`] = {
  "data": "let dir = null;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//a`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//a
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//ab`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//ab
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abc`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abc
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcd`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcd
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcde`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcde
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcdef`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcdef
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcdefg`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`//abcdefg
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`bid 1C!: 16+, artificial`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndeal.bid({level: 1,suit: 'â™£',alert: true,description: `16+, artificial`,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`bid 1C!://
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndeal.bid({level: 1,suit: 'â™£',alert: true,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`bid 1C!:// `] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndeal.bid({level: 1,suit: 'â™£',alert: true,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`bid 1C!:// 
`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndeal.bid({level: 1,suit: 'â™£',alert: true,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`bid P`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndeal.bid({level: 0,});\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer W`] = {
  "data": "let dir = null;\ndeal.dealer = 'west';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer e`] = {
  "data": "let dir = null;\ndeal.dealer = 'east';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer eAst`] = {
  "data": "deal.dealer = 'east';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer n`] = {
  "data": "let dir = null;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer north`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer s`] = {
  "data": "let dir = null;\ndeal.dealer = 'south';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer south`] = {
  "data": "deal.dealer = 'south';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`dealer west`] = {
  "data": "deal.dealer = 'west';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`e`] = {
  "data": "let dir = null;\ndir = deal.east;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`east`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.east;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`format error`] = {
  "data": "Error: Expected \"!\", \"$\", \"%\", \"\u002F\u002F\", \"\u003C\", \"C\", \"D\", \"H\", \"S\", \"\\n\", \"\\r\\n\", \"any\", \"balanced\", \"balanced5\", \"bid\", \"dealer\", \"east\", \"north\", \"singleton\", \"south\", \"void\", \"vuln\", \"west\", [ \\t], [0-9], or end of input but \"f\" found.\n --\u003E testing:1:1\n  |\n1 | foo\n  | ^",
  "meta": {
    "expires": null
  }
}

exports[`n`] = {
  "data": "let dir = null;\ndir = deal.north;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  !!11-13
  4D`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.range(11, 13))) { return false; }\nif (!(((dir.diamonds.length === 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  !22-23, balanced`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (dir.range(22, 23) && dir.balancedNoM()) { return false; }\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  !22-23, balanced5`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (dir.range(22, 23) && dir.balanced()) { return false; }\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  !22-23, balanced5 //
`] = {
  "data": "let dir = null;\ndir = deal.north;\nif (dir.range(22, 23) && dir.balanced()) { return false; }\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  !22-23, balanced5 // precision
`] = {
  "data": "let dir = null;\ndir = deal.north;\nif (dir.range(22, 23) && dir.balanced()) { return false; }\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  $ðž‹„ = H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nlet ðž‹„ = null;\ndir = deal.north;\nðž‹„ = `H`;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  %POINTS = 4, 3, 2, 1
  S%POINTS > 3`] = {
  "data": "let dir = null;\nlet POINTS = null;\ndir = deal.north;\nPOINTS = [4, 3, 2, 1];\nif (!((Deal.weight(dir.spades, POINTS) \u003E 3))) {\n  return false;\n}\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  %POINTS = 40, 3, 2, 1
  S%POINTS > 3`] = {
  "data": "let dir = null;\nlet POINTS = null;\ndir = deal.north;\nPOINTS = [40, 3, 2, 1];\nif (!((Deal.weight(dir.spades, POINTS) \u003E 3))) {\n  return false;\n}\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  %POINTS = 40, 30, 2, 1
  S%POINTS > 3`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nlet POINTS = null;\ndir = deal.north;\nPOINTS = [40, 30, 2, 1];\nif (!((Deal.weight(dir.spades, POINTS) \u003E 3))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  %POINTS = 40, 30, 2, 1
  S%POINTS > 30`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nlet POINTS = null;\ndir = deal.north;\nPOINTS = [40, 30, 2, 1];\nif (!((Deal.weight(dir.spades, POINTS) \u003E 30))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  13`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.points === 13))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  13+`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.points \u003E= 13))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  13-15`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.range(13, 15))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  37`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.points === 37))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4+M`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(((dir.hearts.length \u003E= 4) || (dir.spades.length \u003E= 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4-6S`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length \u003E= 4) && (dir.spades.length \u003C= 6))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4441`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.isShape(4, 4, 4, 1))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4CD`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(((dir.clubs.length === 4) || (dir.diamonds.length === 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4HS`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(((dir.hearts.length === 4) || (dir.spades.length === 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4X`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(((dir.clubs.length === 4) || (dir.diamonds.length === 4) || (dir.hearts.length === 4) || (dir.spades.length === 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  4m`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(((dir.clubs.length === 4) || (dir.diamonds.length === 4)))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  8`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.points === 8))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <5b`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.clubs.length \u003C 5) && (dir.spades.length \u003C 5))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <5r`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.diamonds.length \u003C 5) && (dir.hearts.length \u003C 5))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <6R`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.clubs.length \u003C 6) && (dir.hearts.length \u003C 6))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <6r`] = {
  "data": "let dir = null;\ndir = deal.north;\nif (!((dir.clubs.length \u003C 6) && (dir.hearts.length \u003C 6))) {\n  return false;\n}\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <=6P`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.diamonds.length \u003C= 6) && (dir.spades.length \u003C= 6))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  <=6p`] = {
  "data": "let dir = null;\ndir = deal.north;\nif (!((dir.diamonds.length \u003C= 6) && (dir.spades.length \u003C= 6))) {\n  return false;\n}\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  CAK`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.hasCards('CAK')))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  S < H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length \u003C dir.hearts.length))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  S <= H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length \u003C= dir.hearts.length))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  S = H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length === dir.hearts.length))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  S > H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length \u003E dir.hearts.length))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  S >= H`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!((dir.spades.length \u003E= dir.hearts.length))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  any 4441`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.isShapeAny(4, 4, 4, 1))) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  singleton`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.hasSingleton())) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
  void`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\nif (!(dir.hasVoid())) {\n  return false;\n}\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
$suit = S`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nlet suit = null;\ndir = deal.north;\nsuit = `S`;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
south`] = {
  "data": "let dir = null;\ndir = deal.north;\ndir = deal.south;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`north
south // foo`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.north;\ndir = deal.south;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`s`] = {
  "data": "let dir = null;\ndir = deal.south;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`south`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.south;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`south
%points = 4, 3, 2, 1`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nlet points = null;\ndir = deal.south;\npoints = [4, 3, 2, 1];\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln BOTH`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'All'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln all`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'All'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln ew`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'EW'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln neither`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'None'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln none`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'None'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln ns`] = {
  "data": "deal.dealer = 'north';\ndeal.vuln = 'NS'\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`vuln random`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`w`] = {
  "data": "let dir = null;\ndir = deal.west;\ndeal.dealer = 'north';\ndeal.randVuln();\nreturn true;\n",
  "meta": {
    "expires": null
  }
}

exports[`west`] = {
  "data": "deal.dealer = 'north';\ndeal.randVuln();\nlet dir = null;\ndir = deal.west;\nreturn true;\n",
  "meta": {
    "expires": null
  }
}
